flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/HeapInternalTimerServiceTest.java

```diff
@@ -0,0 +1,509 @@
+import org.mockito.invocation.InvocationOnMock;
+import org.mockito.stubbing.Answer;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.InputStream;
+
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.contains;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.*;
+
+/**
+ * Tests for {@link HeapInternalTimerService}.
+ */
+public class HeapInternalTimerServiceTest {
+
+	private static InternalTimer<Integer, String> anyInternalTimer() {
+		return any();
+	}
+
+	/**
+	 * Verify that we only ever have one processing-time task registered at the
+	 * {@link ProcessingTimeService}.
+	 */
+	@Test
+	public void testOnlySetsOnePhysicalProcessingTimeTimer() throws Exception {
+		@SuppressWarnings("unchecked")
+		Triggerable<Integer, String> mockTriggerable = mock(Triggerable.class);
+
+		TestKeyContext keyContext = new TestKeyContext();
+
+		TestProcessingTimeService processingTimeService = new TestProcessingTimeService();
+
+		HeapInternalTimerService<Integer, String> timerService =
+				createTimerService(mockTriggerable, keyContext, processingTimeService);
+
+		keyContext.setCurrentKey(0);
+
+		timerService.registerProcessingTimeTimer("ciao", 10);
+		timerService.registerProcessingTimeTimer("ciao", 20);
+		timerService.registerProcessingTimeTimer("ciao", 30);
+		timerService.registerProcessingTimeTimer("hello", 10);
+		timerService.registerProcessingTimeTimer("hello", 20);
+
+		assertEquals(5, timerService.numProcessingTimeTimers());
+		assertEquals(2, timerService.numProcessingTimeTimers("hello"));
+		assertEquals(3, timerService.numProcessingTimeTimers("ciao"));
+
+		assertEquals(1, processingTimeService.getNumRegisteredTimers());
+		assertThat(processingTimeService.getRegisteredTimerTimestamps(), containsInAnyOrder(10L));
+
+		processingTimeService.setCurrentTime(10);
+
+		assertEquals(3, timerService.numProcessingTimeTimers());
+		assertEquals(1, timerService.numProcessingTimeTimers("hello"));
+		assertEquals(2, timerService.numProcessingTimeTimers("ciao"));
+
+		assertEquals(1, processingTimeService.getNumRegisteredTimers());
+		assertThat(processingTimeService.getRegisteredTimerTimestamps(), containsInAnyOrder(20L));
+
+		processingTimeService.setCurrentTime(20);
+
+		assertEquals(1, timerService.numProcessingTimeTimers());
+		assertEquals(0, timerService.numProcessingTimeTimers("hello"));
+		assertEquals(1, timerService.numProcessingTimeTimers("ciao"));
+
+		assertEquals(1, processingTimeService.getNumRegisteredTimers());
+		assertThat(processingTimeService.getRegisteredTimerTimestamps(), containsInAnyOrder(30L));
+
+		processingTimeService.setCurrentTime(30);
+
+		assertEquals(0, timerService.numProcessingTimeTimers());
+
+		assertEquals(0, processingTimeService.getNumRegisteredTimers());
+	}
```

Comment:
Could have an additional check that it `getNumRegisteredTimers()` goes again back up to 1 if we start registering a timer again.